### PR TITLE
feat: add test command and templates

### PR DESCRIPTION
**Pull request by cmatsuoka**
The `charmcraft test` command invokes spread to run test suites on
local and remote machines. A pre-defined spread configuration and
multipass back-end are supplied.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

### CODE REVIEW

**Comment by cmatsuoka** [comment]
Rebased on main.

File: charmcraft/application/commands/test.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        parser.add_argument(
+            "spread_args",
+            metavar="spread arguments",
+            nargs=argparse.REMAINDER,
+            help="Arguments to spread",
+        )
+
+        # XXX: Arguments to the test command are passed directly to spread.
+        #      Inject separator in the argument to avoid parsing test command
+        #      arguments. We should implement better support for this case
```
Comment (mr-cal) [inline charmcraft/application/commands/test.py:54]: Might be worth creating a craft-application issue and referencing it here

File: charmcraft/templates/init-kubernetes/spread.yaml.j2 [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  LANGUAGE: "en"
+
+  PROJECT_PATH: /home/spread/proj
+  CRAFT_TEST_LIB_PATH: /home/spread/proj/tests/spread/lib
+
+backends:
+  multipass:
+    type: adhoc
+    allocate: |
+      sleep 0.$RANDOM
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply charmcraft/templates/init-kubernetes/spread.yaml.j2:29]: +1 on this and the purpose of `counter_file` and `instance_num` below

File: charmcraft/templates/init-kubernetes/tests/spread/lib/tools/retry.j2 [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    parser.add_argument(
+        "--env",
+        type=envpair,
+        metavar="KEY=VAL",
+        action="append",
+        default=[],
+        help="environment variable to use with format KEY=VALUE (no default)",
+    )
+    parser.add_argument(
+        "--maxmins",
```
Comment (mr-cal) [inline charmcraft/templates/init-kubernetes/tests/spread/lib/tools/retry.j2:52]: I thought this was 'maxims' at first :smile: 

This is easier to read and follows the same style as `--expect-rc`.
```suggestion
        "--max-mins",
```

File: charmcraft/templates/init-simple/tests/spread/lib/tools/retry.j2 [hunk 4]
```diff
@@ -0,0 +1,158 @@
+#!/usr/bin/env python3
```
Comment (mr-cal) [inline charmcraft/templates/init-simple/tests/spread/lib/tools/retry.j2:1]: Does `init-machine` need this script too?

Are these identical?
- `charmcraft/templates/init-simple/tests/spread/lib/tools/retry.j2`
- charmcraft/templates/init-kubernetes/tests/spread/lib/tools/retry.j2

If so, is it possible to symlink these files?

**Review by mr-cal — COMMENTED** [review]
Looking good!

File: charmcraft/templates/init-kubernetes/spread.yaml.j2 (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmatsuoka) [reply charmcraft/templates/init-kubernetes/spread.yaml.j2:29]: Added a comment and a reference to https://github.com/canonical/multipass/issues/3336

File: charmcraft/application/commands/test.py (hunk 1 above)
Comment (cmatsuoka) [reply charmcraft/application/commands/test.py:54]: Created https://github.com/canonical/craft-application/issues/327

File: charmcraft/templates/init-kubernetes/tests/spread/lib/tools/retry.j2 (hunk 3 above)
Comment (cmatsuoka) [reply charmcraft/templates/init-kubernetes/tests/spread/lib/tools/retry.j2:52]: This file was copied as-is from the parca tests, fixing as suggested.

File: charmcraft/templates/init-simple/tests/spread/lib/tools/retry.j2 (hunk 4 above)
Comment (cmatsuoka) [reply charmcraft/templates/init-simple/tests/spread/lib/tools/retry.j2:1]: It is, good idea.

File: charmcraft/application/commands/test.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            spread_args = spread_args[1:]
+
+        if len(spread_args) == 0 and os.environ.get("GITHUB_RUN_ID"):
+            spread_args = ["github-ci"]
+
+        try:
+            cmd = f"{os.environ['SNAP']}/bin/spread"
+            with emit.pause():
+                subprocess.run([cmd, *spread_args], check=True)
+        except subprocess.CalledProcessError as err:
```
[the remark this answers is not in the document]
Comment (cmatsuoka) [reply charmcraft/application/commands/test.py:79]: Refactored a little bit to check whether running in snap before using the bundled spread binary.

**Review by mr-cal — APPROVED** [review]
Nice work!

Do you need `charmcraft/templates/init-machine/tests/spread/lib/tools/retry.j2`?

File: charmcraft/application/commands/test.py [hunk 6]
```diff
@@ -27,8 +27,8 @@
 _overview = """
 Run charm tests in different back-ends.
 
-This command will run charm test suites using the spread tool. See
-the spread documentation for further information.
+This command will run charm test suites using the spread tool. Forfurther
```
[the remark this answers is not in the document]
Comment (cmatsuoka) [reply charmcraft/application/commands/test.py:31]: oh oops, fixing!

File: charmcraft/application/commands/test.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            nargs=argparse.REMAINDER,
+            help="Arguments to spread",
+        )
+
+        # XXX: Arguments to the test command are passed directly to spread.
+        #      Inject separator in the argument to avoid parsing test command
+        #      arguments. We should implement better support for this case
+        #      in craft-application.
+        idx = sys.argv.index("test") + 1
+        sys.argv = [*sys.argv[:idx], "--", *sys.argv[idx:]]
```
[the remark this answers is not in the document]
Comment (lengau) [reply charmcraft/application/commands/test.py:58]: Yeah, especially with the way this obscures the help text. I think long-term we'll probably need to make our own items for common parameters and then let the user pass `-- <spread-params>` if they want to do anything fancy with spread.

**Review by lengau — APPROVED** [review]

**Pull request merged**